### DISCUSSION TITLE
Is it possible to debug D1 databases and/or durable objects locally?

### DISCUSSION THREAD

**Discussion by endereyewxy** [skymethod/denoflare#78] [Q&A] [answered]
It seems that `denoflare d1` doesn't have the `--local` option like `wrangler`. Will Denoflare simply use remote databases and durable objects when `denoflare serve`ing?

**Comment by johnspurlock-skymethod** [comment, marked as answer]
For durable objects:

In production (running on cloudflare via denoflare push) The `doNamespace` binding value looks like: `my-do-namespace-name:MyDOClass`

```json
"bindings": {
    "backendNamespace": { "doNamespace": "backend-prod:BackendDO" },
}
```
For local serving, you can use a binding value of `local:MyDOClass`, which will instantiate your DO class locally and use an in-memory version for transactional storage (no persistence)

```json
"bindings": {
    "backendNamespace": { "doNamespace": "local:BackendDO" },
}
```

or

you can add a few options to the local backend for persistent storage, `storage=webstorage` and name the local storage container `container=my-container-name`

```json
"bindings": {
    "backendNamespace": { "doNamespace": "local:BackendDO:storage=webstorage:container=mp1-local" },
}
```

D1 support in denoflare is newer, but in all cases you set the `d1DatabaseUuid` binding value to a guid.

```json
"bindings": {
    "db1": { "d1DatabaseUuid": "6a34fd2e-8cc4-4f74-8722-3c4373788793" },
}
```

In production, that specifies the D1 database UUID. See `denoflare d1` for the commands to manage databases.

In development, D1 is backed by a local SQLite database, and the guid defines the container.
